lib/http/response-validator/validationStreams.js

```diff
@@ -0,0 +1,131 @@
+var AWS = require('../../core');
+var cryptoLib = AWS.util.crypto.lib;
+var Buffer = AWS.util.Buffer;
+var TransformStream = AWS.util.stream.Transform;
+var CHECKSUM_BYTE_LEN = 16;
+
+module.exports = {
+  IntegrityCheckerStream: IntegrityCheckerStream,
+  ContentLengthCheckerStream: ContentLengthCheckerStream,
+}
+
+require('util').inherits(IntegrityCheckerStream, TransformStream);
+require('util').inherits(ContentLengthCheckerStream, TransformStream);
+
+/**
+ * Only runnable in Node
+ * A transform stream to perform trailing md5 hash validation.
+ * Whether to perform this validation is based on 'x-amz-tranfer-encoding'
+ * response header.
+ * @api private
+ */
+function IntegrityCheckerStream(transferEncoding, contentLength) {
+  if (!(this instanceof TransformStream)) {
+    return new TransformStream();
+  }
+  TransformStream.call(this);
+  if (transferEncoding !== 'append-md5') {
+    process.nextTick(function(stream, encoding) {
+      stream.emit(
+        'error',
+        AWS.util.error(
+          new Error('Cannot validate response header: ' + encoding + ', expect \'append-md5\''),
+          {code: 'UnsupportedHashingAlgorithm'}
+        )
+      );
+    }, this, transferEncoding);
+  }
```

Comment:
it might be inevitable. Here data may come earlier than error event get thrown and stream unpiped. Then we need to access instantiated objects to do the transform, in which case if we return too early, we may access undefined objects

Reply:
I think it makes sense to just throw the error immediately - since this is a constructor method and the condition we're checking is synchronous - rather than emit an error. This way we'll fail fast and not waste resources. It's also what other node APIs, like fs.createReadStream do if you pass in bad parameters.